pombola/south_africa/views/download.py

```diff
@@ -0,0 +1,86 @@
+import os
+
+from django.db.models import Prefetch, Q
+from django.http import Http404, StreamingHttpResponse
+from django.shortcuts import get_object_or_404
+from django.views.generic import TemplateView
+
+import xlsx_streaming
+from pombola.core.models import Contact, Organisation, Person, Position
+
+MP_DOWNLOAD_TEMPLATE_SHEET = os.path.join(
+    os.path.dirname(os.path.realpath(__file__)), "mp-download-template.xlsx"
+)
+
+
+def get_email_address_for_person(person):
+    if person.email:
+        return person.email
+    if len(person.email_addresses) > 0:
+        return person.email_addresses[0].value
+    return ""
+
+
+def download_members_xlsx(request, slug):
+    organisation = get_object_or_404(Organisation, slug=slug)
+
+    # Get all of the currently_active positions at the organisation
+    organisation_positions = (
+        Position.objects.currently_active()
+        .filter(organisation=organisation)
+        .values("id")
+    )
+
+    party_positions = (
+        Position.objects.currently_active()
+        .filter(title__slug="member")
+        .filter(organisation__kind__slug="party")
+        .select_related("organisation")
+    )
+
+    cell_phone_contacts = Contact.contact_number_contacts()
+    email_contacts = Contact.email_contacts()
+
+    # Get the persons from the positions
+    persons = (
+        Person.objects.filter(hidden=False)
+        .filter(position__id__in=organisation_positions)
+        .distinct()
```

Comment:
I think the person (and related objects) queries should be pulled out from the view method.

That would allow you to test

- do we have all the active MPs
- do we have none of the inactive MPs
- do we have only NA MPs and no NCOP MPs
in dedicated tests, and without a lot of logic traversing an excel file, so you could end up with something like the following tests:

- integration test verifying that we indeed get back an excel file, with values in the right places
- check that all active MPs are included - using the queryset
- check that no inactive MPs are included - using the queryset
- check no NCOP members returned when querying NA members - using the queryset

each test only needs the data relevant to that test, e.g. for the inactive members test, it just needs 1 inactive member.

I think this will be a lot clearer: what data is relevant to what aspect we're trying to ensure.

It also raises the question: isn't there already a query somewhere for "give me the active members of this house"? e.g. for the house members page https://www.pa.org.za/organisation/national-assembly/people/ or the MP contact pombola API https://pa.org.za/api/national-assembly/popolo.json

they'd need the same logic and tests, wouldn't they?

Reply:
> isn't there already a query somewhere for "give me the active members of this house"?

This part is quite trivial because we can just use the relationship: `organisation.position_set.currently_active()`.